### PR TITLE
#1349: Continue of refactoring of `ScalarOf`

### PR DESCRIPTION
**Pull request by andreoss**
Per #1349: 

- Add `ScalarOfSupplier`
- Add `ScalarOfCallable`
- Remove `ScalarOf`

### CODE REVIEW

**Comment by victornoel** [comment]
@0crat assign @fabriciofx

**Comment by 0crat** [comment]
> @0crat assign @fabriciofx  ([here](https://github.com/yegor256/cactoos/issues/1435#issuecomment-687623143))

@victornoel This pull request [#1435](https://github.com/yegor256/cactoos/issues/1435) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), here is [why](https://www.0crat.com/footprint/C63314D6Z/64a021b7-5b51-48c1-af30-28ef4de2943b); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @victornoel[/z](https://www.0crat.com/u/victornoel) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/org/cactoos/scalar/ScalarOfCallableTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.jupiter.api.Test;
+import org.llorllale.cactoos.matchers.Assertion;
+import org.llorllale.cactoos.matchers.ScalarHasValue;
+
+/**
+ * Test case for {@link ScalarOfCallable}.
+ *
+ * @since 0.47
+ */
+final class ScalarOfCallableTest {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/ScalarOfCallableTest.java:36]: @andreoss makes sense add test for exception?

File: src/test/java/org/cactoos/scalar/ScalarOfSupplierTest.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.junit.jupiter.api.Test;
+import org.llorllale.cactoos.matchers.Assertion;
+import org.llorllale.cactoos.matchers.ScalarHasValue;
+
+/**
+ * Test case for {@link ScalarOfSupplier}.
+ * @since 0.47
+ */
+final class ScalarOfSupplierTest {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/ScalarOfSupplierTest.java:34]: @andreoss makes sense add test for exception?

**Review by fabriciofx — CHANGES_REQUESTED** [review]
@andreoss can you look it, please?

File: src/test/java/org/cactoos/scalar/ScalarOfCallableTest.java (hunk 1 above)
Comment (andreoss) [reply src/test/java/org/cactoos/scalar/ScalarOfCallableTest.java:36]: @fabriciofx I would add such test to `ScalarEnvelopeTest` rather, and avoid having two very similar tests

File: src/test/java/org/cactoos/scalar/ScalarOfSupplierTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Test
+    void createsScalarFromSupplier() {
+        final Object obj = new Object();
+        new Assertion<>(
+            "must hold the same value",
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/test/java/org/cactoos/scalar/ScalarOfSupplierTest.java:40]: @fabriciofx Done.

File: src/test/java/org/cactoos/scalar/ScalarOfCallableTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    void worksWithCallable() {
+        final Object obj = new Object();
+        new Assertion<>(
+            "must hold the same value",
```
[the remark this answers is not in the document]
Comment (andreoss) [reply src/test/java/org/cactoos/scalar/ScalarOfCallableTest.java:41]: @fabriciofx Done.

File: src/test/java/org/cactoos/scalar/ScalarOfSupplierTest.java (hunk 2 above)
Comment (andreoss) [reply src/test/java/org/cactoos/scalar/ScalarOfSupplierTest.java:34]: @fabriciofx Added a test in `ScalarEnvelopeTest`

**Comment by andreoss** [comment]
@fabriciofx Please have a look

**Comment by fabriciofx** [comment]
@andreoss it looks fine to me, but now it has a conflict. Can you fix it?

**Comment by andreoss** [comment]
@fabriciofx Conflicts fixed.

File: src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java [hunk 5]
```diff
@@ -45,6 +47,22 @@ public void envelopeDelegatesCalls() {
         ).affirm();
     }
 
+    @Test
+    public void propagatesException() {
+        new Assertion<>(
+            "must not alter the exception thrown by original Scalar",
+            new ScalarEnvelope<Integer>(
+                () -> {
+                    throw new UnsupportedOperationException("ok");
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java:56]: @andreoss I'm sorry, but I haven't see this "ok" twice. Can you move them to a single local variable?

File: src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java [hunk 6]
```diff
@@ -45,6 +47,22 @@ public void envelopeDelegatesCalls() {
...
+                }) {
```
Comment (fabriciofx) [inline src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java:57]: @andreoss Is there many ```{``` here, isn't it? Can you check it and correct the ident?

**Review by fabriciofx — CHANGES_REQUESTED** [review]
@andreoss please, can you check it again? Thanks!

File: src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java (hunk 6 above)
Comment (andreoss) [reply src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java]: @fabriciofx Extracted parameter to variable

File: src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java (hunk 5 above)
Comment (andreoss) [reply src/test/java/org/cactoos/scalar/ScalarEnvelopeTest.java]: @fabriciofx Done.

**Comment by fabriciofx** [comment]
@andreoss it seems fine to me! Thanks!
@victornoel can you merge it, please?

**Review by victornoel — CHANGES_REQUESTED** [review]
@andreoss let's keep `ScalarOf` anyway (taking a `Scalar` and extending `ScalarEnvelope`), even though it seems superfluous. ~~And let's use it in `ThreadsTest` and `TimedTest`, there is no reason to rely on the concept of `Callable` there I believe.~~

**Comment by victornoel** [comment]
@andreoss scratch that, let me think a bit about it, sorry

**Comment by victornoel** [comment]
@andreoss ok, so:
- let's keep `ScalarOf`
- but let's use lambdas directly it in `ThreadsTest` and `TimedTest`, there is no reason to rely on the concept of `Callable` or use `ScalarOf` there I believe.

Btw `ScalarOf` will be improved as part of #1445 that's why we should keep it.

**Comment by andreoss** [comment]
@victornoel `TimedTest` uses `CallableOf` (not `ScalarOfCallable`), but it's unnecessary too. Do you want `CallableOf` removed from this test?

**Comment by victornoel** [comment]
@andreoss yes, that's what I meant, let's remove unneeded code :)

**Comment by andreoss** [comment]
@victornoel Restored `ScalarOf`, removed `CallableOf` from tests

**Comment by victornoel** [comment]
@andreoss thank you for your patience

**Comment by victornoel** [comment]
@rultor merge

**Pull request merged**